File: exercism/kotlin/phone-number/src/main/kotlin/PhoneNumber.kt
class PhoneNumber(val raw: String) {

  val number: String
    get

  init {
    val sanitized: String = raw.filter { it.isDigit() }
    if (sanitized.length == 10 ||
        sanitized.length == 11 && sanitized.startsWith("1")) {
      number = sanitized.takeLast(10)
    } else {
      number = "0000000000"
    }
  }

  val areaCode: String
    get() = number.take(3)

  override fun toString(): String {
    return "($areaCode) ${number.substring(3, 6)}-${number.substring(6, 10)}"
  }
}

File: exercism/elixir/phone-number/phone_number.exs
defmodule Phone do
  @invalid "0000000000"

  @spec number(String.t) :: String.t
  def number(raw) do
    clean = String.replace(raw, ~r/[\(\)\-\. ]/, "")
    cond do
      not String.match?(clean, ~r/^[0-9]+$/) -> @invalid
      String.length(clean) == 10 -> clean
      String.length(clean) == 11 and String.starts_with?(clean, "1")
        -> String.slice(clean, 1, 11)
      true -> @invalid
    end
  end

  @spec area_code(String.t) :: String.t
  def area_code(raw) do
    raw |> number
        |> String.slice(0, 3)
  end

  @spec pretty(String.t) :: String.t
  def pretty(raw) do
    num = number(raw)
    "(#{String.slice(num, 0, 3)}) #{String.slice(num, 3, 3)}-#{String.slice(num, 6, 4)}"
  
  end
end
